File: Data Structures & Algorithms/max-water-container/submission-1.rs
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut left = 0usize;
        let mut right = height.len() - 1;
        let mut best = 0;

        while left < right {
            let width = (right - left) as i32;

            if height[left] <= height[right] {
                best = best.max(height[left] * width);
                left += 1;
            } else {
                best = best.max(height[right] * width);
                right -= 1;
            }
        }

        best
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        n = len(heights)
        left = [0] * n
        right = [n] * n
        stack = []

        for i, height in enumerate(heights):
            while stack and heights[stack[-1]] >= height:
                stack.pop()

            left[i] = stack[-1] + 1 if stack else 0
            stack.append(i)

        stack = []

        for i in range(n - 1, -1, -1):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()

            right[i] = stack[-1] if stack else n
            stack.append(i)

        best = 0

        for i, height in enumerate(heights):
            best = max(best, height * (right[i] - left[i]))

        return best